util: add inspect.defaultOptions

silverwind opened the pull request:
##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
util

##### Description of change
Adds the `util.inspect.config()` method to allow customization of inspect options in cases where `util.inspect()` is implicitly called by core like in `console` or `util.format`.

Fixes: #7566

jasnell commented on doc/api/util.md:
```diff
@@ -207,6 +207,23 @@ Values may supply their own custom `inspect(depth, opts)` functions, when
 called these receive the current `depth` in the recursive inspection, as well as
 the options object passed to `util.inspect()`.
 
+### util.inspect.config([options])
```
Perhaps a `defaultOptions` setter/getter would be better here? Something like...

```js
util.inspect.defaultOptions = options;

// or even...

util.inspect.defaultOptions.maxArrayLength = 1000;
```

silverwind replied:
Right, seems cleaner.

silverwind replied:
Hmm, I think it might be possible to make them plain properties. Would save some time having to compute the getter/setter names at least.

cjihrig commented on lib/util.js:
```diff
@@ -186,19 +195,30 @@ function inspect(obj, opts) {
-  if (ctx.colors === undefined) ctx.colors = false;
-  if (ctx.customInspect === undefined) ctx.customInspect = true;
-  if (ctx.showProxy === undefined) ctx.showProxy = false;
+  Object.keys(inspectOptions).forEach(function(option) {
```
Could you use `Object.assign({}, inspectOptions, ctx)` here?

silverwind replied:
Shouldn't it be `Object.assign(ctx, inspectOptions)`?

cjihrig replied:
Looking at the code, `ctx` contains the user options. Wouldn't `Object.assign(ctx, inspectOptions)` overwrite the user options with the default options?

silverwind replied:
Right, I'm trying `Object.assign(ctx, inspectOptions, opts || {})` now, which eliminates another `_extend` call above.

cjihrig replied:
In that case, yes, that might work. I think you can even drop the `|| {}` part.

jasnell commented:
I'm generally +1 on having this, but prefer making it a getter/setter rather than a method.

cjihrig commented on lib/util.js:
```diff
@@ -186,19 +195,30 @@ function inspect(obj, opts) {
 exports.inspect = inspect;
 
+inspect.config = function (options = {}) {
+  if (typeof options !== 'object') {
```
Shouldn't this check for `null`?

silverwind replied:
Right, `null` is an issue. How about a strict object check instead?

```js
if (Object.prototype.toString.call(options) === '[object Object]') {
```

cjihrig replied:
In other places in core, I believe we've just been doing `options === null || typeof options !== 'object'`

cjihrig commented on lib/util.js:
```diff
@@ -186,19 +195,30 @@ function inspect(obj, opts) {
+    throw new TypeError('"options" argument must be a object');
+  }
+
+  // copy properties that exist on `inspectOptions`
```
Can you capitalize "copy".

cjihrig commented on lib/util.js:
```diff
@@ -186,19 +195,30 @@ function inspect(obj, opts) {
+    }
+  });
+
+  // return a clone of the current options
```
Same with "return" here.

cjihrig commented on lib/util.js:
```diff
@@ -186,19 +195,30 @@ function inspect(obj, opts) {
+  }
+
+  // copy properties that exist on `inspectOptions`
+  Object.keys(options).forEach(function(opt) {
```
You could probably get away with `Object.assign()` or `util._extend()` here. If there are unsupported options in `options`, they'll just be ignored later anyway, right?

silverwind commented:
Thanks for review. What's the general consensus on `util._extend` vs `Object.assign`? Shall we use the former in performance-critical paths, otherwise the latter?

cjihrig commented:
Last I heard, we were still preferring `util._extend()` for performance reasons. However, I believe `Object.assign()` is more flexible, so it might be preferable in cases that would require multiple calls to `util._extend()`.

jbergstroem commented:
Windows fail looks unrelated (Jenkins -- connection aborted)

silverwind commented:
Also, not sure what's up with the linter on CI:

```
+ gmake lint-ci
./node tools/jslint.js -j 1 -f tap -o test-eslint.tap \
	benchmark lib src test tools
gmake: *** [Makefile:677: jslint-ci] Error 1
```

Definitely passes locally.

silverwind commented:
Nevermind, there's some lint errors. Fixing them.

silverwind commented:
Updated to use `util.inspect.defaultOptions` instead of a function.

- I sealed the object to prevent addition/deletion of options
- Currently, the property is `writable: false` to prevent shenanigans like `defaultOptions = null` which would break `inspect`. The downside of this is that we can't set using a object, just through individual properties, one at a time. Thoughts about `writeable`, @jasnell?

jasnell commented:
Using a getter/setter would help protect against things like `defaultOptions = null`, just have the setter function check to see if the input is valid and ignore it if it is not.

I'm not convinced that sealing the object is necessary but it's likely harmless.

silverwind commented:
@jasnell Sealing for example prevents `delete defaultOptions.depth` for of the current solution without accessors.

Supporting both `defaultOptions = {}` and `defaultOptions.property = true` through getters/setters would introduce quite a big amount of code. I'm not sure if it's really that common to set all properties at once. In my case, I'd only like to change `depth` and `maxArrayLength`.

silverwind commented:
@jasnell I've switched to accessors now. The solution turned out shorter than I thought and should satisfy all cases. I'd like to keep the properties sealed to prevent some misuse like deleting.

silverwind commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/3574/

jasnell commented on doc/api/util.md:
```diff
@@ -207,6 +207,25 @@ Values may supply their own custom `inspect(depth, opts)` functions, when
+set to an object containing one or more valid [`util.inspect()`][] options.
+Setting properties directly is also supported.
+
+Returns the current options as an object.
```
s/options/default options

jasnell commented:
LGTM with a nit if CI is green.

cjihrig commented on test/parallel/test-util-inspect.js:
```diff
@@ -723,3 +723,29 @@ checkAlignment(new Map(big_array.map(function(y) { return [y, null]; })));
+  const arr = Array(101);
+  const oldOptions = Object.assign({}, util.inspect.defaultOptions);
+
+  // set single properties
```
Tiny nit, but could you capitalize "set" here and in the comment below.

cjihrig commented:
LGTM pending CI.

silverwind commented:
CI is green except a hanging Windows box. Giving this PR a bit of time for others to weight in.

Trott commented on doc/api/util.md:
```diff
@@ -215,7 +215,7 @@ by `util.inspect`. This is useful for functions like `console.log` or
 Setting properties directly is also supported.
 
-Returns the current options as an object.
+Returns the current default options as an object.
```
Since this is a property and not a function, it doesn't really "return" anything, right? It just *is* a value. I'd be inclined to delete the line entirely. I don't think it actually adds any information. (I also think it's misleading. If you don't set it, it's `undefined` and not an object.)

Trott replied:
(Whoops, the parenthetical is wrong, ignore it. Also, uh, yeah, looks like I'm commenting on an out-of-date diff. SMH. Sorry.)

Trott commented on doc/api/util.md:
```diff
@@ -271,6 +271,25 @@ util.inspect(obj);
+set to an object containing one or more valid [`util.inspect()`][] options.
+Setting properties directly is also supported.
+
+Returns the current default options as an object.
```
It's not a function, so it doesn't "return" anything, right? I would just remove this line.

silverwind replied:
Yeah, it's just a value now. That line is a leftover from when it was a function. I'll remove it.

evanlucas commented:
LGTM with Trott's nit fixed

silverwind commented:
Nit addressed. One more CI: https://ci.nodejs.org/job/node-test-pull-request/3589/

jasnell commented:
still LGTM :-)

silverwind commented:
Noticed a bug where the REPL module called to `inspect` using the legacy form with undefined properties:

```js
util.inspect(obj, undefined, undefined, true)
```

The `undefined` properties were overriding the defaults in the `Object.assign` call. I fixed the issue by not adding `undefined` props to `ctx` and added a test for it. Also made it copy `opts` last, so it matches previous behaviour.

One last CI: https://ci.nodejs.org/job/node-test-pull-request/3598/

jasnell commented:
oh, good catch @silverwind

silverwind commented:
One unrelated CI failure on Windows. Landed in https://github.com/nodejs/node/commit/d9ef3e1fb3461b79c860005707a002557393b1b9.

silverwind commented:
There was a silly typo in the error message. I `--force-with-lease` pushed https://github.com/nodejs/node/commit/1a6a69a8e006d1fd202ea4899714b1974a5540ad to correct it.
